### ISSUE TITLE
Memory leak in SIFT?

### ISSUE THREAD

**Issue by Lezh1k** [hybridgroup/gocv#205]
Hello there. 
I have one really strange problem with SIFT detector. I use it in our service and seems like there is memory leak. 
I use method DetectAndCompute. When I looked at code I found that `defer KeyPoints_Close` is commented. I was able to make new test binding KeyPoints_Close2 and free memory, but it didn't solve my problem. 
I have no clue how to reproduce it without using service. I mean simple application doesn't cause such big memory consumption. 

For example here is some code: 
```
func TestSift(t *testing.T) {
	gray := gocv.IMRead("path_to_image", gocv.IMReadGrayScale)

	sift := contrib.NewSIFT()
	defer sift.Close()

	res := make([]point, 0) //this is used as result in service's code.
	for iter := 0; iter < 50; iter++ {
		mask := gocv.NewMat()
		defer mask.Close()
		k, d := sift.DetectAndCompute(gray, mask)
		defer d.Close()

		for i, v := range k {
			var tmp []float64
			for j := 0; j < d.Cols(); j++ {
				tmp = append(tmp, float64(d.GetFloatAt(i, j)))
			}
			res = append(res, point{keypoint: v, descriptor: tmp})
		}
	}
	utils.ShowImage(gray) //This function just shows image in new window. 
}
```

Similar code is used in my service and I have crashes (because of memory problem). 
Interesting thing - using BRISK (for example) fixes this problem, but using SURF - not.

I wasn't able to find significant difference between BRISK and SIFT bindings, so maybe there is another problem. I have C++ application and it hasn't such problem, so maybe there is some golang specific things. 

I tried to use pprof, but without any luck (But maybe I didn't use it in right way). Valgrind also shows that there is no memory issues.

I just have no clue what's happening there. Top shows, that memory usage increases by ~300Mb after each call.  Does anybody face this problem?

**Comment by tzununbekov** [comment]
It may be useful to have steps to reproduce:
1) `go run main.go`:
https://gist.github.com/tzununbekov/8c9573f75ea5a35d0c3c15b192adbc7e

2) `curl -i -X POST -H "Content-Type: multipart/form-data" -F "img=@<image>" http://localhost:8080/upload`

3) After a couple curl requests `top` (or any other analog) will show over 600Mb of resident memory allocated to this process while heap memory profile (inuse_space, alloc_space) usually shows not more than 50Mb

**Comment by deadprogram** [comment]
Hello, @Lezh1k I think that `defer` is not doing what you think it is when used in the `for` loop. It does not go out of scope inside the loop.

A better implementation would probably be:

```go
func TestSift(t *testing.T) {
	gray := gocv.IMRead("path_to_image", gocv.IMReadGrayScale)

	sift := contrib.NewSIFT()
	defer sift.Close()

	res := make([]point, 0) //this is used as result in service's code.
	for iter := 0; iter < 50; iter++ {
		mask := gocv.NewMat()
		k, d := sift.DetectAndCompute(gray, mask)

		for i, v := range k {
			var tmp []float64
			for j := 0; j < d.Cols(); j++ {
				tmp = append(tmp, float64(d.GetFloatAt(i, j)))
			}
			res = append(res, point{keypoint: v, descriptor: tmp})
		}
                mask.Close()
                d.Close()
	}
	utils.ShowImage(gray) //This function just shows image in new window. 
}
```

**Comment by Lezh1k** [comment]
Yes, it will be better, but won't solve that problem :( I have already tried :) 
I was talking about this : 
```
// Detect keypoints in an image using SIFT.
//
// For further details, please see:
// https://docs.opencv.org/master/d0/d13/classcv_1_1Feature2D.html#aa4e9a7082ec61ebc108806704fbd7887
//
func (d *SIFT) Detect(src gocv.Mat) []gocv.KeyPoint {
	ret := C.SIFT_Detect((C.SIFT)(d.p), C.Mat(src.Ptr()))
	//defer C.KeyPoints_Close(ret)

	return getKeyPoints(ret)
}
```
This commented defer looks like mistake, but fix doesn't help here.

**Comment by deadprogram** [comment]
Hmm, interesting. I have not tried instantiating multiple `SIFT` at the same time. Do you have a C++ example that does this?

**Comment by Lezh1k** [comment]
Not now, sorry. :( Will be available in 12 hours

**Comment by Lezh1k** [comment]
@tzununbekov  provided some code. thank you  ))

**Comment by Lezh1k** [comment]
```

using namespace cv::xfeatures2d;
int main(int argc, char **argv) {
  (void)argc;
  (void)argv;

  typedef cv::Ptr<cv::xfeatures2d::SIFT>* MSIFT;

  MSIFT features = new Ptr<SIFT>(SIFT::create());
  MSIFT extractor = new Ptr<SIFT>(SIFT::create());
  MSIFT computer = new Ptr<SIFT>(SIFT::create());

  // Turn the image into a Mat
  cv::Mat img = cv::imread("/home/lezh1k/test_data/idmatch_data/kg_id_new/hq_newid_pattern.jpg", IMREAD_COLOR);

  for (int i = 0; i < 100; ++i) {
    cv::Mat descriptors;
    std::vector<cv::KeyPoint> lstKeyPoints;
//    // Detect keypoints

    (*features)->detect(img, lstKeyPoints);
    printf("Found %lu keypoints.\n", lstKeyPoints.size());

    // Extract descriptors
    (*extractor)->compute(img, lstKeyPoints, descriptors);
    printf("Found %d descriptors.\n", descriptors.rows);

    (*computer)->detectAndCompute(img, Mat(), lstKeyPoints, descriptors);
    printf("Detect and compute found %lu keypoints.\n\n", lstKeyPoints.size());
  }

  // Let my memory go!
  delete extractor;
  delete features;
  delete computer;

  std::cout << "Press enter to continue ..." << std::endl;
  std::cin.get();
}

```

**Comment by deadprogram** [comment]
The C++ code that you posted does not do the same things (web serving) as what the Go code does.

Perhaps rewriting the C++ code into the equivalente GoCV code would be a good test here to try to isolate?

**Comment by Lezh1k** [comment]
Sure. 
But we tried this one too. I mean we tried to rewrite go code to make it work like this C++ peace. And we got memory leak, but only in service. 
@deadprogram  I've tried to use defer and close like you mentioned :) . Didn't help. I have big memory consumption and it grows up :)

**Comment by deadprogram** [comment]
Sorry I did not fully understand "But we tried this one too. I mean we tried to rewrite go code to make it work like this C++ peace. And we got memory leak, but only in service."

What do you mean "but only in service"?

**Comment by Lezh1k** [comment]
I mean we got memory leak in [this](https://gist.github.com/tzununbekov/8c9573f75ea5a35d0c3c15b192adbc7e) application, but if we run it as test (like above) or just in main function (without work with web/network, just console application) we don't have memory leak.

**Comment by deadprogram** [comment]
That does not sound like an issue that is specific to GoCV itself vs. how you are using it in your application. 

The C++ example you show does 100 SIFT-based calculations serially one after another without any concurrency.

You might want to test your concurrency with SIFT separately from the web-based part of the application. Perhaps the OpenCV SIFT implementation does not work well in this concurrent case.

**Comment by tzununbekov** [comment]
@deadprogram it is looks like the issue was in `getKeyPoints` method of `contrib` package.
This [fix](https://github.com/tzununbekov/gocv/commit/6c3f14599b718c0abc918eb6879cf26f65d1e180) works for described test [case](https://github.com/hybridgroup/gocv/issues/205#issuecomment-391718343), but I'm not sure about its correctness.

**Comment by deadprogram** [comment]
I've made a couple of changes, including the one you have proposed @tzununbekov thanks for that. They will be included in the next GoCV release. I will leave this issue open until that code lands.

**Comment by deadprogram** [comment]
This fix is now released as part of v0.14.0 so closing. Thank you!
